Add x509_crl module

felixfontein opened the pull request:
##### SUMMARY
This module allows to create and/or update Certificate Revocation Lists (CRLs).

I decided to prefix the module name with `x509_` instead of `openssl_` because as #33715. I would like to rename the other modules as well eventually. Right now, I don't mind to change the name of the new module to `openssl_crl` if we prefer to stick to that prefix for now. We just have to make up our mind until Ansible 2.10 Feature Freeze (approx in 4-5 months, if the timeline is similar to previous releases).

Fixes first part of #53789.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
x509_crl

felixfontein committed:
Add x509_crl module.

felixfontein committed:
Add integration tests.

felixfontein committed:
Fix some errors.

felixfontein committed:
Fix inversion.

felixfontein committed:
Compare name instead of tpye.

felixfontein committed:
Fix fail_json() calls.

felixfontein committed:
Work around rename of serial_number attribute for cryptography 1.4.

felixfontein committed:
Don't die for non-cert loading errors.

felixfontein committed:
One more.

felixfontein committed:
Fix function call.

felixfontein commented:
Ok, now it also works with older cryptography versions (1.2.3) and Python 2.6 (at least theoretically).

ctrufan commented on lib/ansible/modules/crypto/x509_crl.py:
```diff
@@ -0,0 +1,789 @@
+
+    next_update:
+        description:
+            - The point in time from which a new CRL will be issued and the client has to check for it.
```
I feel like "client has to check for it" is a little misleading - but it's hard to come up with a good way to frame it, because RFC5280 and how the entire internet uses next update agreed to disagree.

Functionally, next_update serves as a hard expiration date for 99% of clients... but technically, the field just means when a new CRL is expected (context for anyone reading the comment less familiar than you with the RFCs).

Maybe something like
"The absolute latest point in time by which this I(issuer) is expected to have issued another CRL. Many clients will treat a CRL as expired once I(next_update) occurs."?

felixfontein replied:
Ah, that's definitely good to know. I haven't checked the RFC for this, I mainly looked at the [cryptography.io docs](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.CertificateRevocationList.next_update)...

Your formulation sounds good, thanks!

ctrufan commented on lib/ansible/modules/crypto/x509_crl.py:
```diff
@@ -0,0 +1,789 @@
+        suboptions:
+            path:
+                description:
+                    - Serial number of the certificate.
```
- Path to the certificate.
- The serial number and issuer will be extracted from the certificate.

felixfontein replied:
I guess you know which part I copy'n'pasted... :)

ctrufan reviewed (commented):
Didn't have a chance to review the code this weekend, just some of the documentation.

felixfontein committed:
Fixed/improved descriptions.

felixfontein commented on lib/ansible/modules/crypto/x509_crl.py:
```diff
@@ -0,0 +1,791 @@
+                    try:
+                        ext = cert.extensions.get_extension_for_class(x509.AuthorityKeyIdentifier)
+                        if ext.value.authority_cert_issuer is not None:
+                            result['issuer'] = list(ext.value.authority_cert_issuer)
```
I realize now that this is probably incorrect. The `issuer` field should be set to the certificate's issuer, not to the issuer's issuer. I don't know how to convert the certificate's issuer (a [Name](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.Name) object) to a list of [GeneralName](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.GeneralName) objects, though. So no idea how to improve on this. (This already puzzled me about [AuthorityKeyIdentifier](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.AuthorityKeyIdentifier)'s [authority_cert_issuer](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.AuthorityKeyIdentifier.authority_cert_issuer) -- it has the same "conversion problem".)

ctrufan replied:
Most X.509 attribute types use DirectoryName (specifically with UTF8String or PrinteableString - referring to the ASN.1 choices of directory name here, not python/crypto classes), in Python the DirectoryName is an instance of GeneralName.

Edge cases would probably be a nightmare though, technically older or noncompliant certificates could have values that weren't clean utf8/printable strings.

ctrufan replied:
Deleted my comment as being possible incorrect.

felixfontein replied:
Maybe it's best to just remove this functionality and let `issuer` only be entered manually by the user. We can re-add this later if we find out how to do it correctly :)

felixfontein replied:
I've changed the PR accordingly in bef34a7d71f7b25696a1100ffe35d88e3ce8051d

felixfontein commented:
The naming of this module might change; it depends on what we decide in #63801.

felixfontein committed:
Don't read issuer from certificate file.

renard commented:
@felixfontein For what I had check this module works great. Thanks.

However is there a way to update an existing CRL? This would prevent from having a long list of certificates to revoke each time.

TIA.

ffdybuster commented:
@renard you can use `mode: update` (default is `generate`); that will append the certificates to the CRL (assuming they haven't been added already). Other fields (issuer etc.) have to be set every time though.

renard commented:
Ah yes I should peel my eyes :-)

renard commented:
@ffdybuster However what about setting `revokation_date` automatically if `mode` is `update`?

ffdybuster commented:
One possibility would be to make `+0d` the default for `revocation_date` in general. Then the current timestamp will be used - which isn't idempotent, but if you specify `revocation_date` explicitly (and set it to something not relative) it will be.

renard commented:
Yes the fact the task is not idempotent is a real problem since it changes things that should not.

felixfontein commented:
How about a boolean option `ignore_times`, which when enabled ignores timestamps during idempotency checking? That way, if everything else stays the same, the CRL isn't updated. But if you change something (values like issuer, or add certs which weren't there before), the CRL will be updated with the times specified. From how I understand your use-case, that should make it possible to do what you want?

renard commented:
@felixfontein this should do the work. 

This would allow to use something following and keep configuration simple and concise:

```yaml
    
_default_revocation_options: &_default_revocation_options
  reason: cessation_of_operation
  revocation_date: '+0d'

revoked_certificates:
  - path: /path/to/cert1.pem
    <<: *_default_revocation_options
  - path: /path/to/cert2.pem
    <<: *_default_revocation_options
```

The only potential issue I see is if you add a `revocation_date` afterward for a certificate. I guess the `force` option can handle this strange case.

felixfontein committed:
Allow to ignore timestamps.

felixfontein committed:
Default value for revocation_date.

felixfontein committed:
Update tests.

felixfontein commented:
Yep, either use `force: yes`, or stop setting `ignore_timestamps` to `yes` ;-)

I've implemented this, and added a test which shows that this works (https://github.com/ansible/ansible/pull/63435/files#diff-5a650900dcb10635d2dd2d244720ebe5R149-R241, https://github.com/ansible/ansible/pull/63435/files#diff-542a9127ddc7111d91cd5a12b7126f2dR17-R22).

felixfontein commented:
(I guess this PR also needs similar changes as #66384 and #65400, though that can be added in a subsequent PR. I wouldn't mind getting this merged eventually ;) )

renard commented:
If you forget the `ignore_timestamps` something weird happens each time a new revocation line is added. I think nothing should be done if the serial already exists in the CRL.

```
...
    Serial Number: 4F9D179E4D648AD81A60A60C6FDC5DECF30B543F
        Revocation Date: Jan 16 22:42:15 2020 GMT
        CRL entry extensions:
            X509v3 CRL Reason Code:
                Cessation Of Operation
    Serial Number: 3E5A5C4BE99BF838829C25A5F1716E92ABE23C0C
        Revocation Date: Jan 16 22:42:15 2020 GMT
        CRL entry extensions:
            X509v3 CRL Reason Code:
                Cessation Of Operation
    Serial Number: 4F9D179E4D648AD81A60A60C6FDC5DECF30B543F
        Revocation Date: Jan 16 22:42:45 2020 GMT
        CRL entry extensions:
            X509v3 CRL Reason Code:
                Cessation Of Operation
    Serial Number: 3E5A5C4BE99BF838829C25A5F1716E92ABE23C0C
        Revocation Date: Jan 16 22:42:45 2020 GMT
        CRL entry extensions:
            X509v3 CRL Reason Code:
                Cessation Of Operation
...
```

felixfontein committed:
Mention ignore_timestamps in update docs.

felixfontein commented:
It's not weird, it is expected. If you don't specify `ignore_timestamps: no` and don't specify an explicit `revocation_date` or `last_update`, you request the module to set these to the current timestamp.

I've updated the docs for `update` that you might also want to set `ignore_timestamps` if you use it.

felixfontein committed:
Support privatekey_content, and require some options only if state is present.

felixfontein committed:
Allow to pass certificate in directly.

felixfontein committed:
Add tests.

felixfontein committed:
Fix required_if.

felixfontein committed:
Forgot to encode content.

felixfontein commented:
The PR now allows to feed secondary objects in via content (as in #66384).

felixfontein committed:
Forgot to adjust type.

felixfontein committed:
Allow to return CRL's content directly.

felixfontein commented:
Now an equivalent of #65400 is supported as well. I think this is now done.

ready_for_review

felixfontein commented:
@MarkusTeufelberger @Shaps @Spredzy @Xyon @puiterwijk @resmo @ctrufan does anyone of you happen to have a bit time to review this? :)

felixfontein committed:
return_crl_content -> return_content (as in #65400).

felixfontein committed:
Fix elements.

felixfontein committed:
Fix messages.

felixfontein committed:
Use required_one_of and mutually_exclusive instead of doing the checks by hand.

felixfontein committed:
Fix format.

felixfontein committed:
Skip tests on AIX.

felixfontein committed:
Fix typo.

felixfontein commented:
I've prepared a basic x509_info module here: https://github.com/ansible/ansible/compare/devel...felixfontein:x509_crl_info I'll rebase and create a PR once this is merged.

MarkusTeufelberger commented:
get_relative_time_option(...) might be a good candidate for factoring out into module_utils/crypto.py later.

Other than that I looked through the code, documentation and the tests - since it seems to be good for @fawaf I guess it should be merged.

shipit

felixfontein commented:
@MarkusTeufelberger thanks a lot for reviewing this! :)
